aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -735,7 +735,7 @@ Tensor& matrix_rank_impl(
 
   Tensor tol = at::max(atol.unsqueeze(-1), rtol.unsqueeze(-1) * max_S);
 
-  result = at::sum_out(result, S > tol, /*dim=*/-1);
+  result = at::sum(S > tol, /*dim=*/-1);
```

Comment:
`matrix_rank_impl` was not composite compliant, I'm not sure if this is the correct approach @kshitij12345

Reply:
Usually we update this such that only TensorSubclasses take this compliant path so that we don't have perf regression on usual tensors.

Eg.
https://github.com/pytorch/pytorch/blob/72502b94f37e0df2b1d4c34363cdbf1176bbfe1b/aten/src/ATen/native/Sorting.cpp#L301-L310

However, what concerns me is why this CompositeCompliance issue wasn't caught by an OpInfo based CompositeCompliance tests in `test_ops.py`? Do you have any idea? 

https://github.com/pytorch/pytorch/blob/72502b94f37e0df2b1d4c34363cdbf1176bbfe1b/torch/testing/_internal/opinfo/definitions/linalg.py#L1910-L1965


Thanks!

Reply:
It has CompositeCompliance, there are some comments like `For Composite Compliance, allocate `result` of correct shape to ...`

Here's the error that I am getting with this pr when I move the  don't change this file:

```
FAILED test/functorch/test_vmap.py::TestVmapOperatorsOpInfoCPU::test_op_has_batch_rule_linalg_matrix_rank_cpu_float32 - RuntimeError: Batching rule not implemented for aten::sum.IntList_out; the fallback path doesn't work on out= or view ops.
```

Should I instead create a batch rule for `aten::sum.IntList_out`?

Reply:
IIRC, using `out` variant is CompositeCompliant but `out` doesn't work with `vmap`. That is why it was composite compliant. The reason it worked with vmap was because it had a custom batch rule.

So using `sum` without `out` would be the correct fix. Let's just guard it with `if (isTensorSubclassLike(S))` to take this slow path otherwise use the `out` path to avoid perf regression.

Snippet showing out is CompositeCompliant:
```python
import torch

from torch.testing._internal.composite_compliance import generate_cct_and_mode
from torch.testing import make_tensor

CCT, mode = generate_cct_and_mode()

t = make_tensor((5, 5, 5), device='cpu', dtype=torch.float, noncontiguous=False)
c_t = CCT(t, mode)

def fn(t):
    out = t.new_empty(t.shape)
    return torch.sin(t, out=out)

fn(c_t)

```